cub/cub/device/dispatch/tuning/tuning_unique_by_key.cuh

```diff
@@ -344,364 +320,273 @@ template <class KeyT,
+  using delay_constructor                            = detail::no_delay_constructor_t<1155>;
 };
 
-} // namespace unique_by_key
-} // namespace detail
-
-template <typename KeyInputIteratorT, typename ValueInputIteratorT = unsigned long long int*>
-struct DeviceUniqueByKeyPolicy
+template <class KeyT, class ValueT>
+struct policy_hub
 {
-  using KeyT   = typename std::iterator_traits<KeyInputIteratorT>::value_type;
-  using ValueT = typename std::iterator_traits<ValueInputIteratorT>::value_type;
+  // using KeyT   = typename std::iterator_traits<KeyInputIteratorT>::value_type;
+  // using ValueT = typename std::iterator_traits<ValueInputIteratorT>::value_type;
 
-  // SM350
-  struct Policy350 : ChainedPolicy<350, Policy350, Policy350>
+  template <int NOMINAL_4B_ITEMS_PER_THREAD, int THREADS>
+  struct DefaultPolicy
   {
-    static constexpr int INPUT_SIZE = sizeof(KeyT);
-    enum
-    {
-      NOMINAL_4B_ITEMS_PER_THREAD = 9,
-      ITEMS_PER_THREAD            = Nominal4BItemsToItems<KeyT>(NOMINAL_4B_ITEMS_PER_THREAD),
-    };
-
+    static constexpr int nominal_4b_items_per_thread = NOMINAL_4B_ITEMS_PER_THREAD;
+    static constexpr int ITEMS_PER_THREAD            = Nominal4BItemsToItems<KeyT>(nominal_4b_items_per_thread);
     using UniqueByKeyPolicyT =
-      AgentUniqueByKeyPolicy<128,
+      AgentUniqueByKeyPolicy<THREADS,
                              ITEMS_PER_THREAD,
                              cub::BLOCK_LOAD_WARP_TRANSPOSE,
                              cub::LOAD_LDG,
                              cub::BLOCK_SCAN_WARP_SCANS,
                              detail::default_delay_constructor_t<int>>;
   };
 
-  struct DefaultTuning
-  {
-    static constexpr int INPUT_SIZE = sizeof(KeyT);
-    enum
-    {
-      NOMINAL_4B_ITEMS_PER_THREAD = 11,
-      ITEMS_PER_THREAD            = Nominal4BItemsToItems<KeyT>(NOMINAL_4B_ITEMS_PER_THREAD),
-    };
+  struct Policy350
+      : DefaultPolicy<9, 128>
+      , ChainedPolicy<350, Policy350, Policy350>
+  {};
 
-    using UniqueByKeyPolicyT =
-      AgentUniqueByKeyPolicy<64,
-                             ITEMS_PER_THREAD,
-                             cub::BLOCK_LOAD_WARP_TRANSPOSE,
-                             cub::LOAD_LDG,
-                             cub::BLOCK_SCAN_WARP_SCANS,
-                             detail::default_delay_constructor_t<int>>;
-  };
+  // Use values from tuning if a specialization exists, otherwise pick the default
+  template <typename Tuning>
+  static auto select_agent_policy(int)
+    -> AgentUniqueByKeyPolicy<Tuning::threads,
+                              Tuning::items,
+                              Tuning::load_algorithm,
+                              Tuning::load_modifier,
+                              BLOCK_SCAN_WARP_SCANS,
+                              typename Tuning::delay_constructor>;
+  template <typename Tuning>
+  static auto select_agent_policy(long) -> typename DefaultPolicy<Tuning::items, Tuning::threads>::UniqueByKeyPolicyT;
```

Comment:
Important: If the above overload SFINAEs out, then `Tuning` does not have members, so you cannot access any here.
```suggestion
  static auto select_agent_policy(long) -> typename DefaultPolicy<11, 64>::UniqueByKeyPolicyT;
```

Reply:
thanks, fixed